flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/binary/BinaryStringDataUtil.java

```diff
@@ -168,21 +179,35 @@ public static BinaryStringData hash(BinaryStringData str, String algorithm)
     }
 
     /**
-     * Parses this BinaryStringData to DecimalData.
+     * Parses a {@link BinaryStringData} to {@link DecimalData}.
      *
-     * @return DecimalData value if the parsing was successful, or null if overflow
-     * @throws NumberFormatException if the parsing failed.
+     * @return DecimalData value if the parsing was successful.
+     * @throws TableException if the parsing failed or if overflow happened.
      */
-    public static DecimalData toDecimal(BinaryStringData str, int precision, int scale) {
+    public static DecimalData toDecimal(BinaryStringData str, int precision, int scale)
+            throws TableException {
         str.ensureMaterialized();
 
-        if (DecimalDataUtils.isByteArrayDecimal(precision)
-                || DecimalDataUtils.isByteArrayDecimal(str.getSizeInBytes())) {
-            return toBigPrecisionDecimal(str, precision, scale);
+        DecimalData data;
+
+        try {
+            if (DecimalDataUtils.isByteArrayDecimal(precision)
+                    || DecimalDataUtils.isByteArrayDecimal(str.getSizeInBytes())) {
+                data = toBigPrecisionDecimal(str, precision, scale);
+            } else {
+                int sizeInBytes = str.getSizeInBytes();
+                data =
+                        toDecimalFromBytes(
+                                precision, scale, getTmpBytes(str, sizeInBytes), 0, sizeInBytes);
+            }
+        } catch (NumberFormatException e) {
+            throw new TableException("Cannot parse string to decimal", e);
```

Comment:
Probably it makes sense to specify what string cannot be parsed like it is done for `toBoolean`.